### PR TITLE
Exit calls in build scripts no longer build to abort

### LINKED ISSUE #20020: Pkg.build has silent failure when exit() used within build.jl
When installing "Coverage" in a 32-bit Linux environment something strange happens:
```julia
julia> Pkg.add("Coverage")
INFO: Cloning cache of BinDeps from https://github.com/JuliaLang/BinDeps.jl.git
INFO: Cloning cache of BufferedStreams from https://github.com/BioJulia/BufferedStreams.jl.git
INFO: Cloning cache of Codecs from https://github.com/dcjones/Codecs.jl.git
INFO: Cloning cache of Compat from https://github.com/JuliaLang/Compat.jl.git
INFO: Cloning cache of Coverage from https://github.com/JuliaCI/Coverage.jl.git
INFO: Cloning cache of Git from https://github.com/JuliaPackaging/Git.jl.git
INFO: Cloning cache of HttpCommon from https://github.com/JuliaWeb/HttpCommon.jl.git
INFO: Cloning cache of HttpParser from https://github.com/JuliaWeb/HttpParser.jl.git
INFO: Cloning cache of JSON from https://github.com/JuliaIO/JSON.jl.git
INFO: Cloning cache of Libz from https://github.com/BioJulia/Libz.jl.git
INFO: Cloning cache of MbedTLS from https://github.com/JuliaWeb/MbedTLS.jl.git
INFO: Cloning cache of Requests from https://github.com/JuliaWeb/Requests.jl.git
INFO: Cloning cache of SHA from https://github.com/staticfloat/SHA.jl.git
INFO: Cloning cache of URIParser from https://github.com/JuliaWeb/URIParser.jl.git
INFO: Installing BinDeps v0.4.5
INFO: Installing BufferedStreams v0.2.1
INFO: Installing Codecs v0.2.0
INFO: Installing Compat v0.11.0
INFO: Installing Coverage v0.3.2
INFO: Installing Git v0.1.0
INFO: Installing HttpCommon v0.2.6
INFO: Installing HttpParser v0.2.0
INFO: Installing JSON v0.8.0
INFO: Installing Libz v0.2.2
INFO: Installing MbedTLS v0.4.2
INFO: Installing Requests v0.3.12
INFO: Installing SHA v0.3.0
INFO: Installing URIParser v0.1.7
INFO: Building MbedTLS
Using system libraries...
INFO: Package database updated

julia> using Coverage
ERROR: LoadError: HttpParser not properly installed. Please run
Pkg.build("HttpParser")
 in error(::String, ::String, ::Vararg{Any,N}) at ./error.jl:22
 in include_from_node1(::String) at ./loading.jl:488
 in macro expansion; at ./none:2 [inlined]
 in anonymous at ./<missing>:?
 in eval(::Module, ::Any) at ./boot.jl:234
 in process_options(::Base.JLOptions) at ./client.jl:239
 in _start() at ./client.jl:318
```

The problem is that when building `MbedTLS` an [`exit()` is encountered](https://github.com/JuliaWeb/MbedTLS.jl/blob/d77b6d1ed576f99ea3adbcbcec2c359d692038bf/deps/build.jl#L35) which terminates the Julia subprocess which performs the building. Since the subprocess exits normally this causes the installation to appear to completely successfully.

Now the `exit()` call can be removed from the MbedTLS build script but it should probably be addressed that any `build.jl` file could potentially call exit which can mess up the build process.

### PR DESCRIPTION
**Pull request by omus**
Captures and "ignores" `exit` calls made within `build.jl` scripts.

Closes #20020

### CODE REVIEW

**Comment by yuyichao** [comment]
Continuing in `atexit` sounds like a bad idea. Especially as a hack to workaround package bug.

**Comment by omus** [comment]
No disagreements about it being a hack. That said I don't think that a package bug should be allowed to interrupt the build process in base.

**Comment by yuyichao** [comment]
> That said I don't think that a package bug should be allowed to interrupt the build process in base.

You can do that no matter what.

**Comment by omus** [comment]
As @tkelman mentioned https://github.com/JuliaLang/julia/issues/20020#issuecomment-272872726

> Looks like this was an unforeseen consequence of the difference between #13499 vs #13506, where at the time I was concerned about Julia process startup time on Windows. Since that's fixed now and much less of a concern, does seem like it would be safer to do one process per package.

I can update the PR to spawn multiple processes if that is preferable.

**Comment by omus** [comment]
Revised the PR to spawn a new Julia process for each package build script. I should probably do some additional work on how the exceptions are being captured.

**Comment by tkelman** [comment]
This is still a worthwhile change, if you or anyone else has capacity to get it passing.

**Comment by omus** [comment]
Resolved conflicts.

**Comment by tkelman** [comment]
Nice, looks like it's working. This should theoretically not be user-visible, but will make Pkg.build a bit more robust to packages influencing one another. cc @stevengj if you're able to take a look

**Comment by stevengj** [comment]
Why is `julia` spawning time no longer an issue on Windows?

**Comment by tkelman** [comment]
https://github.com/JuliaLang/julia/pull/17179

**Review by stevengj — APPROVED** [review]

**Comment by tkelman** [comment]
@omus would the intermediate commits pass here or should this be squashed?

**Comment by omus** [comment]
All intermediate commits pass locally.

File: base/pkg/entry.jl [hunk 1]
```diff
@@ -609,42 +592,54 @@ function build!(pkgs::Vector, errs::Dict, seen::Set=Set())
...
-                                    --eval $code`), stderr=STDERR), "w", STDOUT)
+    cmd = ```
+        $(Base.julia_cmd()) -O0
+        --compilecache=$(Bool(Base.JLOptions().use_compilecache) ? "yes" : "no")
+        --history-file=no
+        --color=$(Base.have_color ? "yes" : "no")
+        --eval $code
+    ```
+
+    success(pipeline(cmd, stderr=STDERR))
```
Comment (iblis17) [inline base/pkg/entry.jl]: why we do not resend stdout to `STDOUT`?

Comment (iblis17) [reply base/pkg/entry.jl]: MXNet.jl ran into this: https://github.com/dmlc/MXNet.jl/pull/265#issuecomment-327451376

We need to build libmxnet via `make`. The main build step is requires more than 10 mins on travis-ci. But stdout is gone in Julia 0.6+, so the build got killed.

Comment (omus) [reply base/pkg/entry.jl]: Definitely an oversight. We should add a test for this.

**Pull request merged**